Linux consumption cold start improvements. Part - 2

balag0 opened the pull request:
Part 1 - https://github.com/Azure/azure-functions-host/pull/6408/files
This change addresses the following improvements based on profiling

1. Metrics publisher invokes expensive Process apis to collect memory metrics during cold start. Deferring this by 2 seconds moves it out of critical path.
2. Some of the operations in specialization code path were not being jit-ed since the warmup specialization call doesn't invoke all the operations. This change decrypts the host assignment context, validates and downloads a fake Website-Run-From-Pkg url to jit HttpDownload calls. The fake url points to a local endpoint and the downloaded contents are ignored.

Fixes https://github.com/Azure/azure-functions-host/issues/6407

<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

resolves https://github.com/Azure/azure-functions-host/issues/6407

### Pull request checklist

* [X] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [X] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [X] Otherwise: Backport tracked by issue/PR https://github.com/Azure/azure-functions-host/issues/6407
* [X] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

balag0 commented:
Client side changes to support the full fake warm up call.
https://github.com/Azure/azure-functions-docker-private/pull/89

safihamid commented on src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs:
```diff
@@ -251,8 +251,8 @@ public void Initialize()
             Initialize();
-            _processMonitorTimer = new Timer(OnProcessMonitorTimer, null, TimeSpan.Zero, _memorySnapshotInterval);
-            _metricsPublisherTimer = new Timer(OnFunctionMetricsPublishTimer, null, TimeSpan.Zero, _metricPublishInterval);
+            _processMonitorTimer = new Timer(OnProcessMonitorTimer, null, TimeSpan.FromSeconds(2), _memorySnapshotInterval);
```
What's the impact if this is changed to 5 sec? some cold start scenarios might still take longer than 2 sec for now

balag0 replied:
Taking this offline since some of this has dependency on other infrastructure components.

balag0 replied:
Decided to keep it this way since making it larger doesn't seem to make much of a difference. We can revisit if this becomes an issue later.

divyagandhisethi replied:
Yes it's ok to delay this call if needed.

safihamid commented on src/WebJobs.Script.WebHost/Management/InstanceManager.cs:
```diff
@@ -89,16 +90,19 @@ public async Task<string> SpecializeMSISidecar(HostAssignmentContext context, bo
+                Task.Run(async () => await Download(context.GetRunFromPkgContext()));
                 return true;
             }
             else if (_assignmentContext == null)
```
The 26 msec JIT time in the profile comes from running the else part during specialization. anything we can do here?
Or maybe we could manually add this to PGO file and see if it goes away from the profile?

System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult Microsoft.Azure.WebJobs.Script.WebHost.Management.InstanceManager+<>c__DisplayClass11_0+<<StartAssignment>b__1>d].MoveNext(class System.Threading.Thread)

balag0 replied:
Thanks. I tried to invoke only the safe/sideeffect free api calls without touching any other components so it doesnt interfere with the actual cold start call later on.

the else part additionally includes
applying appsettings
switching from standby mode to specialized mode

Both of which can make the container dirty and so we might need a rollback for the fake calls sideeffects.

Yes, Lets try the PGO approach for this section to see if that helps and we could decide if we want to handle this in the warmup call or PGO.

mathewc commented on src/WebJobs.Script.WebHost/Management/InstanceManager.cs:
```diff
@@ -372,25 +375,25 @@ private async Task<string> Download(RunFromPackageContext pkgContext)
                 try
                 {
-                    using (_metricsLogger.LatencyEvent(MetricEventNames.LinuxContainerSpecializationZipDownload))
+                    using (_metricsLogger.LatencyEvent(isWarmupRequest ? MetricEventNames.LinuxContainerSpecializationZipDownloadWarmup : MetricEventNames.LinuxContainerSpecializationZipDownload))
```
nit: maybe hoist this one and the below to a local like you did in ValidateBlobPackageContext

mathewc commented on src/WebJobs.Script.WebHost/Management/InstanceManager.cs:
```diff
@@ -403,31 +406,33 @@ private async Task HttpClientDownload(string filePath, Uri zipUri)
-            (string stdout, string stderr, int exitCode) = RunBashCommand($"aria2c --allow-overwrite -x12 -d {directory} -o {fileName} '{zipUri}'", MetricEventNames.LinuxContainerSpecializationZipDownload);
+            (string stdout, string stderr, int exitCode) = RunBashCommand(
+                $"aria2c --allow-overwrite -x12 -d {directory} -o {fileName} '{zipUri}'",
+                isWarmupRequest ? MetricEventNames.LinuxContainerSpecializationZipDownloadWarmup : MetricEventNames.LinuxContainerSpecializationZipDownload);
```
here too

mathewc commented on src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs:
```diff
@@ -251,8 +251,8 @@ public void Initialize()
             Initialize();
-            _processMonitorTimer = new Timer(OnProcessMonitorTimer, null, TimeSpan.Zero, _memorySnapshotInterval);
-            _metricsPublisherTimer = new Timer(OnFunctionMetricsPublishTimer, null, TimeSpan.Zero, _metricPublishInterval);
+            _processMonitorTimer = new Timer(OnProcessMonitorTimer, null, TimeSpan.FromSeconds(2), _memorySnapshotInterval);
```
Hmmm, I wonder why the other ProcessMonitor we start immediately isn't showing up in your analysis (code [here](https://github.com/Azure/azure-functions-host/blob/826c0b520c1e81d2f3106d08711c08f1bce4aa95/src/WebJobs.Script/Scale/HostPerformanceManager.cs#L48)). This guy also starts immediately and starts taking CPU samples. There's another one started for the OOP worker.

mathewc replied:
Also, what will the effect be of us having no memory snapshots taken before functions start executing? We don't know the memory usage for the first 2 seconds of invocations.

balag0 replied:
Looks like that process monitor starts before specialization unlike the metrics one which happens in the middle of specialization. So it probably doesnt show up prominently in the cold start profile.

Yes, this can lead to some imprecision in the metering but it should self correct quickly so the impact should be minimal. Checked with Divya also on this to make sure we are okay doing this.

mathewc commented on src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs:
```diff
@@ -251,8 +251,8 @@ public void Initialize()
-            _processMonitorTimer = new Timer(OnProcessMonitorTimer, null, TimeSpan.Zero, _memorySnapshotInterval);
-            _metricsPublisherTimer = new Timer(OnFunctionMetricsPublishTimer, null, TimeSpan.Zero, _metricPublishInterval);
+            _processMonitorTimer = new Timer(OnProcessMonitorTimer, null, TimeSpan.FromSeconds(2), _memorySnapshotInterval);
+            _metricsPublisherTimer = new Timer(OnFunctionMetricsPublishTimer, null, TimeSpan.FromSeconds(2), _metricPublishInterval);
```
nit: recommend defining a _timerStartDelay member for these, peer to the interval definitions

mathewc commented on src/WebJobs.Script.WebHost/Models/HostAssignmentContext.cs:
```diff
@@ -36,6 +35,13 @@ public class HostAssignmentContext
+        // returns 200 to ensure downloading app contents succeeds.
+        // All the other fields will be empty.
+        [JsonProperty("isWarmupRequest")]
+        public bool IsWarmupRequest { get; set; }
```
Can we get into a case in the rollout of this where this value isn't being sent yet to an updated host, meaning it'll try to do a full assign?

balag0 replied:
Good question, there are 2 clients that makes this call, the mesh service and antares.
We want all requests from Antares to always do the full assign. Antares will never set EncryptedHostAssignmentContext.WarmUp or HostAssignmentContext.IsWarmupRequest. So none of the requests from antares will be treated as a warmup request with or without this PR.

For the other client, yes as soon as this PR is merged the corresponding PR in docker should also be merged. I called this out in the linked PR so we dont forget this
https://github.com/Azure/azure-functions-docker-private/pull/89#issuecomment-668799586

In case we miss this it will completely brick the image since no site assignment can take place and so it will fail in an obvious that will fail our OGFs and sanity tests.

mathewc commented on test/WebJobs.Script.Tests.Integration/ContainerManagement/LinuxContainerInitializationHostServiceTests.cs:
```diff
@@ -79,12 +79,12 @@ public async Task Assigns_Context_From_CONTAINER_START_CONTEXT()
             _environment.SetEnvironmentVariable(ContainerEncryptionKey, containerEncryptionKey);
             AddLinuxConsumptionSettings(_environment);
 
-            _instanceManagerMock.Setup(manager => manager.StartAssignment(It.Is<HostAssignmentContext>(context => hostAssignmentContext.Equals(context)), It.Is<bool>(w => !w))).Returns(true);
```
rather than remove these checks, can you move them into the preceeding It.Is check to verify !context.IsWarmupRequest?

balag0 committed:
Linux consumption cold start improvements. Part - 2

    Part 1 - https://github.com/Azure/azure-functions-host/pull/6408/files
    This change addresses the following improvments based on profiling

    1. Metrics publisher invokes expensive Process apis to collect memory metrics during cold start. Deferring this by 2 seconds moves it out of critical path.
    2. Some of the operations in specialization code path were not being jit-ed since the warmup specialization call doesn't invoke all the operations. This change decrypts the host assignment context, validates and downloads a fake Website-Run-From-Pkg url to jit HttpDownload calls. The fake url points to a local endpoint and the downloaded contents are ignored.

    Fixes https://github.com/Azure/azure-functions-host/issues/6407
